File: kotlin/lib/library/src/androidTest/java/com/xplpc/library/TodoTest.kt
package com.xplpc.library

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.xplpc.client.Client
import com.xplpc.core.Config
import com.xplpc.core.XPLPC
import com.xplpc.message.Param
import com.xplpc.message.Request
import com.xplpc.serializer.JsonSerializer
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class TodoTest {
    data class Todo(
        val id: Long,
        val title: String,
        val body: String,
        val data: Map<String, String>,
        val done: Boolean,
    )

    @Before
    fun setUp() {
        XPLPC.initialize(
            Config(JsonSerializer()),
        )
    }

    @Test
    fun singleItem() {
        val todo = Todo(
            1,
            "Title 1",
            "Body 1",
            mapOf("data1" to "value1", "data2" to "value2"),
            true
        )

        val request = Request("sample.todo.single", Param("item", todo))

        Client.call<Todo>(request) { response ->
            assertEquals("Title 1", response?.title)
        }
    }

    @Test
    fun singleItemAsync() {
        val todo = Todo(
            1,
            "Title 1",
            "Body 1",
            mapOf("data1" to "value1", "data2" to "value2"),
            true
        )

        val request = Request("sample.todo.single", Param("item", todo))

        runBlocking {
            Client.call<Todo>(request) { response ->
                assertEquals("Title 1", response?.title)
            }
        }
    }

    @Test
    fun multipleItems() {
        val todo1 = Todo(
            1,
            "Title 1",
            "Body 1",
            mapOf("data1" to "value1", "data2" to "value2"),
            true
        )

        val todo2 = Todo(
            2,
            "Title 2",
            "Body 2",
            mapOf("data1" to "value1", "data2" to "value2"),
            true
        )

        val request = Request("sample.todo.list", Param("items", listOf(todo1, todo2)))

        Client.call<List<Todo>>(request) { response ->
            assertEquals(2, response?.size)
            assertEquals("Title 1", response?.get(0)?.title)
            assertEquals("Title 2", response?.get(1)?.title)
        }
    }

    @Test
    fun multipleItemsAsync() {
        val todo1 = Todo(
            1,
            "Title 1",
            "Body 1",
            mapOf("data1" to "value1", "data2" to "value2"),
            true
        )

        val todo2 = Todo(
            2,
            "Title 2",
            "Body 2",
            mapOf("data1" to "value1", "data2" to "value2"),
            true
        )

        val request = Request("sample.todo.list", Param("items", listOf(todo1, todo2)))

        runBlocking {
            Client.call<List<Todo>>(request) { response ->
                assertEquals(2, response?.size)
                assertEquals("Title 1", response?.get(0)?.title)
                assertEquals("Title 2", response?.get(1)?.title)
            }
        }
    }
}

File: cxx/tests/src/TodoTest.cpp
#include "xplpc/custom/Todo.hpp"
#include "fixtures/GeneralTest.hpp"
#include "xplpc/xplpc.hpp"
#include "gtest/gtest.h"

#include <thread>
#include <vector>

using namespace xplpc::core;
using namespace xplpc::client;
using namespace xplpc::message;

TEST_F(GeneralTest, TodoTestSingle)
{
    auto item = Todo{1, "Todo 1", "Body 1", {}, true};

    auto request = Request{"sample.todo.single", Param<Todo>{"item", item}};

    // clang-format off
    Client::call<Todo>(request, [](const auto &response) {
        EXPECT_NE(response, std::nullopt);

        if (response)
        {
            auto todo = response.value();

            EXPECT_EQ(1, todo.id);
            EXPECT_EQ("Todo 1", todo.title);
            EXPECT_EQ(true, todo.done);
        }
    });
    // clang-format on
}

TEST_F(GeneralTest, TodoTestSingleAsync)
{
    auto item = Todo{1, "Todo 1", "Body 1", {}, true};

    auto request = Request{"sample.todo.single", Param<Todo>{"item", item}};

    // clang-format off
    std::thread([=] {
        Client::call<Todo>(request, [](const auto &response) {
            EXPECT_NE(response, std::nullopt);

            if (response)
            {
                auto todo = response.value();

                EXPECT_EQ(1, todo.id);
                EXPECT_EQ("Todo 1", todo.title);
                EXPECT_EQ(true, todo.done);
            }
        });
    }).join();
    // clang-format on
}

TEST_F(GeneralTest, TodoMultipleTest)
{
    auto item1 = Todo{1, "Todo 1", "Body 1", {}, true};
    auto item2 = Todo{2, "Todo 2", "Body 2", {}, true};
    auto items = std::vector<Todo>{item1, item2};

    auto request = Request{"sample.todo.list", Param<std::vector<Todo>>{"items", items}};

    // clang-format off
    Client::call<std::vector<Todo>>(request, [](const auto &response) {
        auto total = 0;

        if (response)
        {
            auto list = response.value();
            total = list.size();
        }

        EXPECT_EQ(2, total);
    });
    // clang-format on
}

TEST_F(GeneralTest, TodoMultipleTestAsync)
{
    auto item1 = Todo{1, "Todo 1", "Body 1", {}, true};
    auto item2 = Todo{2, "Todo 2", "Body 2", {}, true};
    auto items = std::vector<Todo>{item1, item2};

    auto request = Request{"sample.todo.list", Param<std::vector<Todo>>{"items", items}};

    // clang-format off
    std::thread([=] {
        Client::call<std::vector<Todo>>(request, [](const auto &response) {
            auto total = 0;

            if (response)
            {
                auto list = response.value();
                total = list.size();
            }

            EXPECT_EQ(2, total);
        });
    }).join();
    // clang-format on
}
